lib/logstash/outputs/dynatrace.rb

```diff
@@ -73,12 +73,39 @@ def headers
       def multi_receive(events)
         return if events.length.zero?
 
-        request = Net::HTTP::Post.new(uri, headers)
-        request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
-        response = @client.request(request)
-        return if response.is_a? Net::HTTPSuccess
+        retries = 0
+        begin
+          request = Net::HTTP::Post.new(uri, headers)
+          request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
+          response = send(request)
+          return if response.is_a? Net::HTTPSuccess
+
+          failure_message = "Dynatrace returned #{response.code} #{response.message}."
+
+          if response.is_a? Net::HTTPServerError
+            raise RetryableError.new failure_message
+          end
+
+          if response.is_a? Net::HTTPClientError
+            @logger.error("#{failure_message} Please check that log ingest is enabled and your API token has the `logs.ingest` (Ingest Logs) scope.")
```

Comment:
For the error message you could distinguish between 400 (invalid input, e.g., violated constraints in the payload which could come from an exceeded limit), 404 (which is returned when the endpoint is disabled or permissions are missing) or other 4xx.